artichoke-backend/src/extn/core/env/env_object.rs

```diff
@@ -0,0 +1,295 @@
+    unsafe extern "C" fn env_to_h(mrb: *mut sys::mrb_state, _slf: sys::mrb_value)
+        -> sys::mrb_value;
+}
+
+pub struct Env<T: EnvBackend> {
+    phantom: PhantomData<T>,
+}
+
+impl<T: EnvBackend> Env<T> {
+    const TWO_STRINGS_ARGS_SPEC: &'static [u8] = b"SS!\0";
+
+    unsafe fn extract_two_string_args(interp: &Artichoke) -> (String, Option<String>) {
+        let mut key = <mem::MaybeUninit<sys::mrb_value>>::uninit();
+        let mut value = <mem::MaybeUninit<sys::mrb_value>>::uninit();
+        let mrb: *mut mrb_state = interp.borrow().mrb;
+        sys::mrb_get_args(
+            mrb,
+            Self::TWO_STRINGS_ARGS_SPEC.as_ptr() as *const i8,
+            key.as_mut_ptr(),
+            value.as_mut_ptr(),
+        );
+        let key = key.assume_init();
+        let value = value.assume_init();
+
+        let key_v = Value::new(interp, key);
+        let value_v = Value::new(interp, value);
+
+        if sys::mrb_sys_value_is_nil(value_v.inner()) {
+            (key_v.to_s(), None)
+        } else {
+            (key_v.to_s(), Some(value_v.to_s()))
+        }
+    }
+
+    // set_env
+    // This function may panic if key is empty, contains an ASCII equals sign '='
+    //      or the NUL character '\0', or when the value contains the NUL character.
+    fn validate_set_args(key: &str, value: &Option<String>) -> Result<(), EnvError> {
+        if key.find('=').is_some() || key.find('\0').is_some() {
+            return Err(EnvError::InvalidSetKey);
+        }
+
+        if value.is_some() && value.clone().unwrap().find('\0').is_some() {
+            return Err(EnvError::InvalidSetValue);
+        }
+
+        Ok(())
+    }
+    fn set_internal(key: &str, value: Option<&String>) {
+        T::set_value(key, value);
+    }
+
+    fn string_to_value(interp: &Artichoke, key: &str) -> Value {
+        let gc_was_enabled = interp.disable_gc();
+
+        let result = Value::convert(interp, key);
+
+        if gc_was_enabled {
+            interp.enable_gc();
+        }
+
+        result
+    }
+
+    fn hash_map_to_value(interp: &Artichoke, data: HashMap<String, String>) -> Value {
+        let gc_was_enabled = interp.disable_gc();
+
+        let result = Value::convert(interp, data);
+
+        if gc_was_enabled {
+            interp.enable_gc();
+        }
+
+        result
+    }
+
+    const STRING_SINGLE_ARG_SPEC: &'static [u8] = b"S\0";
+
+    unsafe fn extract_string_arg(interp: &Artichoke) -> Option<String> {
+        let mut other = <mem::MaybeUninit<sys::mrb_value>>::uninit();
+        let mrb: *mut mrb_state = interp.borrow().mrb;
+        sys::mrb_get_args(
+            mrb,
+            Self::STRING_SINGLE_ARG_SPEC.as_ptr() as *const i8,
+            other.as_mut_ptr(),
+        );
+        let other = other.assume_init();
+
+        let arg_value = Value::new(interp, other);
+
+        Some(arg_value.to_s())
+    }
+
+    unsafe fn get_internal(interp: &Artichoke, arg_name: &str) -> sys::mrb_value {
```

Comment:
this function should not be necessary since we can just call `T::get_value` and use a converter.

```rust
Value::convert(interp, T::get_value(arg_name)).inner()
```